Failure cluster [76ae0294...]: TestAuthorizerMetrics/canceled_request

### Failure cluster [76ae0294f168d11c2f9a](https://go.k8s.io/triage#76ae0294f168d11c2f9a)

##### Error text:
```
Failed;Failed;
=== RUN   TestAuthorizerMetrics/canceled_request
    metrics_test.go:140: incorrect status code recorded for RecordRequestTotal method, expected = <error>, got 200
    metrics_test.go:144: incorrect status code recorded for RecordRequestLatency method, expected = <error>, got 200
    metrics_test.go:148: expected evaluationsResult "canceled", got "success"
    metrics_test.go:151: expected durationResult "canceled", got "success"
    --- FAIL: TestAuthorizerMetrics/canceled_request (0.07s)
;=== RUN   TestAuthorizerMetrics
--- FAIL: TestAuthorizerMetrics (0.47s)
;
```
#### Recent failures:
[20/3/2025, 1:41:02 AM ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit/1902452527295631360)
[18/3/2025, 6:37:30 PM ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit/1901983606029422592)
[16/3/2025, 10:50:26 PM ci-kubernetes-unit-1-32](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-1-32/1901322493193359360)
[16/3/2025, 11:50:30 AM ci-kubernetes-unit-1-31](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-1-31/1901156397261787136)
[16/3/2025, 7:34:36 AM ci-kubernetes-unit-1-30](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-1-30/1901091972060286976)
https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-ppc64le/1904867966478979072

/kind failing-test
/kind flake
/sig auth
<!-- Please assign a SIG using: /sig SIG-NAME -->

--- a/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/metrics_test.go
+++ b/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/metrics_test.go
@@ -107,9 +107,16 @@ func TestAuthorizerMetrics(t *testing.T) {
 			if service.statusCode == 0 {
 				service.statusCode = 200
 			}
-			service.reviewHook = func(*authorizationv1.SubjectAccessReview) {
-				if scenario.canceledRequest {
+
+			testFinishedCtx, testFinishedCancel := context.WithCancel(context.Background())
+			defer testFinishedCancel()
+			if scenario.canceledRequest {
+				service.reviewHook = func(*authorizationv1.SubjectAccessReview) {
 					cancel()
+					// net/http transport still attempts to use a response if it's
+					// available right when it's handling context cancellation,
+					// we need to delay the response.
+					<-testFinishedCtx.Done()
 				}
 			}
 			service.allow = !scenario.authFakeServiceDeny
@@ -120,6 +127,9 @@ func TestAuthorizerMetrics(t *testing.T) {
 				return
 			}
 			defer server.Close()
+			// testFinishedCtx must be cancelled before we close the server, otherwise
+			// closing the server hangs on an active connection from the listener.
+			defer testFinishedCancel()
 
 			fakeAuthzMetrics := &fakeAuthorizerMetrics{}
 			wh, err := newV1Authorizer(server.URL, scenario.clientCert, scenario.clientKey, scenario.clientCA, 0, fakeAuthzMetrics, cel.NewDefaultCompiler(), []apiserver.WebhookMatchCondition{}, "")

